runtime/compiler/trj9/optimizer/ValueJProfiling.cpp

```diff
@@ -0,0 +1,883 @@
+      if (origNode && check->getChild(i) == origNode)
+         {
+         check->setAndIncChild(i, newNode);
+         origNode->decReferenceCount();
+         }
+      else
+         replaceNode(check->getChild(i), origNode, newNode, checklist);
+      }
+   }
+
+/*
+ * Replace all references to origNode under and after replaceStart with newNode.
+ *
+ * \param blockStart Start of the extended block containing replaceStart.
+ * \param origNode Node to replace.
+ * \param newNode Node to use in its place.
+ */
+void
+TR_ValueJProfiling::replaceNode(TR::Compilation *comp, TR::TreeTop *blockStart, TR::TreeTop *replaceStart,
+    TR::Node *origNode, TR::Node *newNode)
+   {
+   TR::NodeChecklist checklist(comp);
+   TR::TreeTop *tt = blockStart; 
+
+   // Collect all nodes seen before start of replace, these shouldn't be replaced
+   while (tt != replaceStart)
+      {
+      replaceNode(tt->getNode(), NULL, NULL, checklist);
+      tt = tt->getNextTreeTop();
+      }
+
+   // Begin replacing until the end of the extended block
+   while (tt)
+      {
+      if (tt->getNode()->getOpCodeValue() == TR::BBStart && !tt->getNode()->getBlock()->isExtensionOfPreviousBlock())
+         break;
+      replaceNode(tt->getNode(), origNode, newNode, checklist);
+      tt = tt->getNextTreeTop();
+      }
+   }
+
+/*
+ * Mark a node and its children as profiling code.
+ *
+ * \param node Node to process, marking it and its children.
+ * \param checklist Checklist of nodes already marked.
+ */
+void
+TR_ValueJProfiling::setProfilingCode(TR::Node *node, TR::NodeChecklist &checklist)
+   {
+   if (checklist.contains(node))
+      return;
+   checklist.add(node);
+
+   node->setIsProfilingCode();
+   for (uint32_t i = 0; i < node->getNumChildren(); ++i)
+      setProfilingCode(node->getChild(i), checklist);
+   }
+
+/*
+ * Generate the tree to store a value into a temporary symref. If no symref is specified, one will be generated.
+ *
+ * \param value Value to store.
+ * \param symRef Optional symref to store value into. Will be update to used symref if none specified.
+ */
+TR::Node *
+TR_ValueJProfiling::storeNode(TR::Compilation *comp, TR::Node *value, TR::SymbolReference* &symRef)
+   {
+   if (symRef == NULL)
+      symRef = comp->getSymRefTab()->createTemporary(comp->getMethodSymbol(), value->getDataType());
+
+   return TR::Node::createWithSymRef(value, directStore(value->getDataType()), 1, value, symRef);
+   }
+
+/*
+ * Given an address in X86's LEA form, generate a node structure to calculate a memory address for a desired
+ * data type. Address would be: base + width(dataType) * index + offset
+ *
+ * \param dataType Date type of value to address. Its width is used for calculations if index is specified.
+ * \param base Base address to load from.
+ * \param index Optional index.
+ * \param offset Optional offset from the base.
+ */
+TR::Node *
+TR_ValueJProfiling::effectiveAddress(TR::DataType dataType, TR::Node *base, TR::Node *index, TR::Node *offset)
```

Comment:
Seems oddly x86 specific. Will this code run for other platforms as well?

Reply:
I will let @ncough comment, but I don't think this style is a problem on other platforms - it exposes maximum richness and the optimizer will have an opportunity to rearrange if necessary as will the code generators.